qmetaobject_impl/src/doc.rs

```diff
@@ -0,0 +1,100 @@
+/* Copyright (C) 2020 ivan tkachenko a.k.a. ratijas <me@ratijas.tk>
+
+Permission is hereby granted, free of charge, to any person obtaining a copy of this software and
+associated documentation files (the "Software"), to deal in the Software without restriction,
+including without limitation the rights to use, copy, modify, merge, publish, distribute, sublicense,
+and/or sell copies of the Software, and to permit persons to whom the Software is furnished to do so,
+subject to the following conditions:
+
+The above copyright notice and this permission notice shall be included in all copies or substantial
+portions of the Software.
+
+THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR IMPLIED, INCLUDING BUT
+NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND
+NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES
+OR OTHER LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN
+CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.
+*/
+
+use darling::FromMeta;
+use proc_macro::TokenStream;
+use syn::{AttributeArgs};
+
+#[derive(Default, FromMeta)]
+#[darling(default)]
+pub struct QtDocMetaArgs {
+    qt: String,
+    // exactly one of them must be present at a time.
+    cls: Option<String>,
+    method: Option<String>,
+    member: Option<String>,
+    related: Option<String>,
+    typedef: Option<String>,
+}
+
+fn count<T>(it: &Option<T>) -> usize {
+    if it.is_some() { 1 } else { 0 }
+}
+
+impl QtDocMetaArgs {
+    fn count_optionals(&self) -> usize {
+        0
+            + count(&self.cls)
+            + count(&self.method)
+            + count(&self.member)
+            + count(&self.related)
+            + count(&self.typedef)
+    }
+
+    fn kind_and_item(&self) -> (&str, &str) {
+        if self.count_optionals() != 1 {
+            panic!(r#"Exactly one of attributes must be specified: cls, method, member, ..."#);
```

Comment:
It would be better to return an Error  so that syn can generate a compile_error! out of it.

(I know that this crate panics because this wasn't supported before)

Reply:
OK. I will address error reporting in later PRs but before lending this branch onto master.

After all this macro is all about what _can be done_ with it, rather that what have to be denied. So, graceful recovering from misuses has lower priority, I think.

In the long term outlook, this should become public API, and allow 3rd-party developers wrap parts of Qt modules following upcoming guidelines per #81.